### ISSUE TITLE
Does any training framework know glm5_next?

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/dgx-manager#67]
Part of #65

## Question

**Does any training framework support `glm5_next` at all?** This is the most likely hard blocker in the map — if nothing can build a trainable `Glm5NextForConditionalGeneration`, the venue question is moot in both venues.

The asymmetry that makes this urgent: vLLM shipped GLM-5.3-Flash support **day-0**, including an arm64 image we already run. Training stacks routinely lag new model classes by months, and the class was not in vLLM 0.28.0 *release* either — we run a special tag.

Establish:

1. **`transformers`.** Is `Glm5NextForConditionalGeneration` in a released version, in `main`, or in an open PR only? Which version first carries it? Does it come with a working `AutoProcessor` / image processor, or text-only? Is there a `remote_code` path in the `zai-org` repo that works for training as well as inference?
2. **PEFT.** Does LoRA work over whatever `transformers` exposes — in particular, do the target modules resolve, and does PEFT handle the MoE expert stack and the router sanely (or does `all-linear` catastrophically target 288 experts)?
3. **The trainer ecosystem.** LLaMA-Factory, ms-swift, Axolotl, Unsloth, torchtune, NeMo — does any of them list GLM-5.3-Flash or `glm5_next` as supported? ms-swift and LLaMA-Factory historically add GLM models fast; check their model registries and recent commits, not just their READMEs.
4. **The multimodal path specifically.** Even where the text model trains, does the image path work — processor, `pixel_values`, and whether the chat template renders image turns correctly for training (not just inference).
5. **Blockers with dates.** If support is in-flight, name the PR/issue and its state. If a fork is the only route, name it and say what it is against.

**Answer shape.** A support matrix (framework × text-training / vision-training / LoRA / status), then a single recommendation: which stack, at which version/commit, is the least-bad route — or an explicit "nothing works today, and here is what would have to land".

**Note.** Report what is true as of the research date and say so; this is a fast-moving surface and the answer has a short half-life.

**Comment by kreuzhofer** [comment]
## Answer: yes — exactly one framework, and it is not one on the ticket's list

**Research date 2026-08-31.** Full write-up:
[`docs/research/glm5next-training-support.md`](https://github.com/kreuzhofer/dgx-manager/blob/research/glm5next-training-support/docs/research/glm5next-training-support.md)
(branch `research/glm5next-training-support`).

**The feared hard blocker did not materialise.** `transformers` 5.16.1 carries a *training-shaped*
implementation — not a thin inference port — and **NVIDIA-NeMo/Automodel merged native GLM-5.3-Flash
training support on 2026-08-28**, two days after the model shipped, with a numerically-validated
VLM SFT recipe. The venue question is live, not moot.

### Support matrix

| Stack | Text training | Vision training | LoRA | Status / blocker |
|---|:---:|:---:|:---:|---|
| **NeMo Automodel** ≥ `9228f33c` | **✅ validated** | **✅ validated** (image; not video) | **⚠️ untested here** | Native `glm5_next` impl + published recipe. `GroupedExpertsLoRA` and `glm_5.x_lora.yaml` exist, but **no `glm5_next` LoRA recipe**. No TP/PP/single-GPU. |
| `transformers` 5.16.1 + `peft` 0.20.0 + `trl` | ✅ wired, untested | ✅ wired, untested | ⚠️ partial | All hooks present. But `all-linear` reaches **2.49%** and mis-targets; experts cost 1.42 B trainable @ r=8 + ~13.5 GiB/layer; **packing impossible**; no FlashAttention. |
| `transformers` ≤ **5.16.0** | ❌ | ❌ | ❌ | Class does not exist. It is in **v5.16.1 only** — merged 24 min before that patch release. |
| **Axolotl** `0.19.0.dev0` | ⚠️ generic fallback | ⚠️ generic fallback | ⚠️ generic | Pins `transformers==5.16.1` and its multimodal allowlist is *derived* from transformers, so the class loads — but **zero** `glm5_next` refs, no processing strategy, no template. Its GLM-5.2 DSA kernels do not carry over. |
| **Unsloth** (CUDA) | ❌ | ❌ | ❌ | Pins `transformers<=5.5.0`. Cannot construct the class. |
| **Unsloth** (MLX) | ✅ | ✅ | ✅ | Real, merged unsloth-zoo#1120 (2026-08-28). Irrelevant to this fleet. |
| **LLaMA-Factory** | ❌ | ❌ | ❌ **silently wrong** | Zero matches. Pins `transformers<=5.8.0`, re-checked at runtime. Never supported **any** GLM-5.x. Unregistered multimodal types **no-op instead of raising** — `--freeze_vision_tower` silently ignored, LoRA attaches to the vision tower. |
| **ms-swift** | ❌ **fail-fast** | ❌ | ❌ | Zero matches. Registers `GlmMoeDsaForCausalLM` (5/5.1/5.2), a text-only arch. Generic fallback **raises** for multimodal configs. Pin (`<5.17.0`) is fine; only registration is missing. |
| **torchtune** | ❌ | ❌ | ❌ | **Deprecated 2025-07-15.** No GLM of any generation. No `transformers` dependency at all. |
| **Megatron-Bridge / NeMo-RL** | ❌ | ❌ | ❌ | Zero matches; both pin `transformers<=5.12.1`. |
| *(context)* **vLLM `main`** | — | — | — | `Glm5NextForConditionalGeneration` is **not in vLLM `main`** either. |

### Recommendation

**NVIDIA-NeMo/Automodel at ≥ `9228f33c` (PR [#3699](https://github.com/NVIDIA-NeMo/Automodel/pull/3699),
merged 2026-08-28T07:27:29Z), starting from
`examples/vlm_finetune/glm5_next/glm5_3_flash_medpix_packed2k_ep72_cp2_100steps.yaml` plus a `peft:`
block adapted from `examples/llm_finetune/glm/glm_5.2_lora.yaml`.**

That recipe already ships four of the five things the map decided it wanted: real image+text training
on `zai-org/GLM-5.3-Flash`, `freeze_vision_tower: true` as a flag, boundary-correct sequence packing,
and proper 288-expert EP. The fifth — parameter-efficient instead of full FT — is the only piece not
composed, but both halves exist: `FinetuneRecipeForVLM` already reads a `peft:` block
(`recipes/vlm/finetune.py` L507-509, L537-538), and `glm_5.2_lora.yaml` shows the GLM-MoE target
syntax including `"*.mlp.experts"`, backed by `GroupedExpertsLoRA` — which injects into the
grouped-GEMM expert forward instead of reconstructing the stack. **Composing them is a YAML change,
not a code change — and it is the highest-value probe this map can run next.**

It does not need `transformers` 5.16: NeMo pins `5.12.1` and owns the implementation. It does need
FlashMLA built from source for the cuDNN backend (or `backend.attn: sdpa`, −13.3% throughput).

**Fallback** if NeMo's operational cost is too high: `transformers==5.16.1` + `peft>=0.20.0` + `trl`,
with the explicit `target_modules` list in §4.5 of the write-up. Never `"all-linear"`.

### Per sub-question

1. **`transformers`** — released, in **v5.16.1 only** ([#48342](https://github.com/huggingface/transformers/pull/48342),
   merged 2026-08-26T14:26:41Z; `v5.16.0` 404s for that path, `v5.16.1` published 24 min after the
   merge). Full `AutoProcessor` path: `Glm5NextProcessor` + `Glm5NextImageProcessor(TorchvisionBackend)`
   + video processor, all native. **There is no remote-code path at all** — the HF repo has no
   `modeling_*.py` — so the "works for inference, breaks for training" failure mode does not apply.
   Training hooks are all present: `supports_gradient_checkpointing = True`, `GradientCheckpointingLayer`
   decoder + vision blocks, `labels → loss_function`, MoE `load_balancing_loss_func`, `pixel_values`
   on `forward`, and the inherited `test_training_gradient_checkpointing*` tests are **unskipped**.
2. **PEFT** — no `glm5_next` entry in `constants.py`, and `total_count: 0` issues mentioning it. **The
   question was backwards:** the 288 routed experts are 3-D `nn.Parameter` stacks, so `all-linear`
   **cannot see them at all** — it reaches **8.01 B of 321.32 B = 2.49%**, and among what it *does*
   reach are the vision tower the map froze and a DSA indexer whose `forward` is `@torch.no_grad()`
   (dead adapters). The router is safe by accident. Reaching the experts needs `target_parameters`
   (peft 0.20.0), which costs **1 420 M trainable params at r=8** (vs 10.5 M for #49's Qwen3.8 LoRA)
   and a **13.50 GiB `baddbmm` materialisation per MoE layer forward**. `lora_dropout` must be 0.
3. **Trainer ecosystem** — see matrix. Registries grepped, not READMEs. Note `NVIDIA/NeMo` now
   redirects to `NVIDIA-NeMo/Speech` (no LLM collection); `LLaMA-Factory` → `hiyouga/LlamaFactory`;
   ms-swift's `swift/llm/` tree is gone. A global code search for `glm5_next` finds **NeMo Automodel
   as the only training framework** among ~1 028 matches.
4. **Multimodal** — processor ✅, `pixel_values` reaches the loss ✅, the chat template **does** render
   image turns correctly (`emit_image()` → `<|begin_of_image|><|image|><|end_of_image|>`, expanded by
   `replace_image_token`). **But the template has no `{% generation %}` block**, so
   `return_assistant_tokens_mask=True` cannot work and **loss masking must be hand-built** — straight
   onto #55. Also: it is a reasoning model whose template emits `<think></think>` when no
   `reasoning_content` is present, so every training target needs an explicit think segment or you
   teach it that an empty reasoning block is the right prefix (cf. the Qwen3.8 reasoning-effort bug).
5. **Blockers with dates** — nothing that would add support to a framework currently lacking it is
   in flight. Both `glm5_next` training PRs (NeMo #3699, unsloth-zoo #1120) are **merged**; the only
   open ones are unsloth-zoo #1126 (GGUF export) and #1137 (MLX test hardening). LLaMA-Factory,
   ms-swift, Axolotl and torchtune have **zero** open issues or PRs mentioning it.

### Two of the map's other open questions fall out

- **Does the MTP head survive a tune?** `transformers` sets
  `_keys_to_ignore_on_load_unexpected = [r"layers\.45\.", ...]` and its model doc states outright
  *"The implementation in transformers does not include an MTP layer"*; NeMo's recipe sets
  `num_nextn_predict_layers: 0`. So **base + adapter served on the vLLM side image keeps MTP intact**,
  but any `save_pretrained()` round-trip (merge, re-quantise) **drops it** along with the 1.57–1.82×
  decode. Merge → copy layer 45's shards by hand.
- **Venue.** Three independent verified grounds rule out the Sparks: `zai-org/GLM-5.3-Flash-BF16` is
  **321.32 B BF16 = 598.5 GiB of weights** vs 498 GB pooled; EP must divide 288, so EP4 on four Sparks
  = **155.8 GB of expert weights per 124.5 GB GB10** before optimizer state (NeMo's validated
  topologies are EP72/72×H100 at 57.68 GiB/GPU and EP144/144×H100); and there is **no FLA /
  `causal_conv1d` build for sm_121** — transformers' own kernel table says so — so 34 of 45 layers
  would train through a pure-PyTorch chunked Python loop.

### Also worth knowing

- **`AutoModelForCausalLM` will not load this model.** `glm5_next` is registered only in
  `MODEL_FOR_IMAGE_TEXT_TO_TEXT_MAPPING_NAMES`; there is no `Glm5NextForCausalLM`. This breaks
  `dgx-manager-fine-tune-recipes/train.py` as written, even for a text-only tune.
- **Sequence packing is impossible under `transformers`** — its own tests skip the padding-free cases
  with *"Fundamentally incompatible with indexer — indexer has no boundary offset telling sequences
  apart."* NeMo solved this (`packing_format: neat`, FLA carrying recurrent state across shards),
  which is one of the strongest arguments for that route.
- **The vendor's "Fine-tuning GLM-5 Series Models" README section naming ms-swift is stale** — added
  2026-07-09, seven weeks before GLM-5.3-Flash existed, and untouched by the three GLM-5.3 update
  commits. It is a claim about the dense-MoE GLM-5 line, not about `glm5_next`.
- **The asymmetry is stranger than assumed:** the arm64 serving image
  `vllm/vllm-openai:glm53-flash-arm64-cu130` was pushed **2026-08-26 14:10:08 UTC — 16 minutes before
  the model class existed in transformers `main`** — and `Glm5NextForConditionalGeneration` is *still*
  not in vLLM `main` today.

**Comment by kreuzhofer** [comment]
## Correction (2026-09-06): the kernel ground does not hold

Re-verified against primary sources. Full write-up: [`docs/research/sm121-fla-blocker-recheck.md`](https://github.com/kreuzhofer/dgx-manager/blob/research/sm121-fla-blocker-recheck/docs/research/sm121-fla-blocker-recheck.md) (branch `research/sm121-fla-blocker-recheck`).

**The support matrix above stands. The "Sparks are out on kernel grounds" conclusion does not.** All three legs fail:

**1. The cuDNN "SM90 or later" reading was a misreading.** The gate is literal source, `nemo_automodel/components/models/common/cudnn_sparse_attention.py`:

```python
major, minor = torch.cuda.get_device_capability(device)
if major < 9:
    raise RuntimeError(f"{operation} requires SM90 or later, got SM{major}{minor}.")
```

sm_121 has `major=12`; `12 < 9` is False; **it passes**. The claim that "the version numbers do not order the way you'd guess" is wrong — NVIDIA's own code orders them exactly that way, and cuDNN's support matrix lists compute capability 12.1 explicitly, naming SM121 ("Blackwell Consumer") with an SM80 floor for fused SDPA.

The scoping was also wrong: that gate applies to `backend.attn: cudnn` for the **11 DSA layers**, a separate doc bullet from "KDA layers use FLA kernels". It never applied to the 34 KDA layers it was cited against. The cuDNN path *is* unavailable on GB10, but because FlashMLA compiles only `sm_90a`/`sm_100f` — and NeMo's own A/B prices the `sdpa` fallback at **13.26% throughput**. A 13% accelerator was promoted to a hard blocker.

**2. "No `causal_conv1d` build for sm_121" is factually false.** Its `setup.py` has compiled `sm_121` since commit `3d19ec779b`, merged **2025-08-29** — PR #71, *"[NVIDIA] Add support Thor, **Spark** and GB300"*. A year before this ticket. Confirmed empirically rather than from docs: NVIDIA's prebuilt `causal_conv1d-1.5.3-cp312-cp312-linux_aarch64.whl` contains **3 `sm_121` cubins** among sm_75/80/87/90/100/103/110/120/121.

**3. "No FLA build" is a category error.** FLA is pure Triton — `setup.py` builds no extension, PyPI ships only `py3-none-any`, and `fla/ops/kda/` is nine `.py` files. There are no per-arch builds for **any** architecture. FLA also does not depend on Dao's `causal_conv1d` (its own conv defaults to `backend='triton'`), and `fla/utils/_device.py` classes 12.1 as Blackwell, naming GB10 in a comment.

**The `transformers` quote is real but was read backwards.** It exists verbatim at `hub_kernels.py:160` — as a *parenthetical justifying a workaround*. The surrounding code **is** the workaround: an SM121-gated `Atlas-Inference/gdn` Hub kernel merged 2026-06-19, validated on real GB10 hardware at cosine 0.99997 with identical greedy tokens. The documented behaviour is "falls back", never raises.

**Real cost of the fallback: slow, not impossible.** `>1 order of magnitude` vs H100 per the transformers source; a real Qwen3.8-27B SFT measured **25–30 s/it → 169–200 s/it**; "3–5x slower per training step" on B200/B300, and it **converges**. FLA issue #913 is a first-hand DGX Spark report from 2026-06-10: 0.5.0 crashes in the GDN backward, **0.4.2 runs clean** — someone was already training through FLA on a Spark.

### What this does not change

**The venue answer is unaffected.** GLM-5.3-Flash still goes to a datacentre GPU, on the scale grounds #65 already records and this re-check re-confirmed from `config.json`: 598.5 GiB BF16 against ~498 GB pooled, expert parallelism must divide 288, and NeMo states verbatim *"TP and PP are not supported for this model"* and *"Full-model single-GPU checkpoint loading and training are not supported"*, validated only at 9 nodes / 72 H100.

### Two caveats that cut against the Sparks

- The Atlas SM121 entry covers only Qwen3.5/3.6 GDN. **There is no SM121-gated `chunk_kda` entry**, so GLM-5.3-Flash specifically gets no fast path on a Spark.
- **Nobody anywhere has tested FLA's KDA path on sm_121.** That is the exact question this whole argument turned on, and it remains unmeasured. It needs hardware, not more reading.

### Why this matters beyond this ticket

The generalisation this finding produced — *"if it has linear-attention/SSM layers, the answer is no on sm_121 regardless of parameter count or quantization"* — is **withdrawn**. It would wrongly veto training a small GDN/KDA model on the Sparks, which is a real capability. Size and parallelism strategy are the right first questions, not attention architecture.
